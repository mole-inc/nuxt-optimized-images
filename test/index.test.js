const path = require('upath')
const cloneDeepWith = require('lodash.clonedeepwith')
const isString = require('lodash.isstring')

const nuxtOptimizedImages = require('../lib')
const nuxtMock = require('./nuxt.mock')

// Get relative path from current file
const relative = dir => {
  return path.relative(__dirname, dir)
}

// Convert absolute paths to relative
const convartPathsToRelative = (option) => {
  return cloneDeepWith(option, value => {
    if (isString(value) && path.isAbsolute(value)) {
      const relativePath = path.normalize(relative(value))

      if (relativePath.split('../').length <= 3) {
        return path.toUnix(path.normalize(relative(value)))
      }
    }
  })
}

describe('nuxt-optimized-images', () => {
  it('returns a mock nuxt.js config object', () => {
    const nuxtMockObject = nuxtMock()

    expect(typeof nuxtMockObject.init).toBe('function')
    expect(typeof nuxtMockObject.extendBuild).toBe('function')
    expect(typeof nuxtMockObject.options).toBe('object')
  })

  it('handles all images by default', () => {
    const nuxtMockObject = nuxtMock()
    const options = nuxtMockObject.init(nuxtOptimizedImages)

    expect(options.webPackConfig.module.rules).toHaveLength(3)

    const rule = options.webPackConfig.module.rules[0]
    const webpRule = options.webPackConfig.module.rules[1]

    expect(rule.test.test('.jpg')).toEqual(true)
    expect(rule.test.test('.jpeg')).toEqual(true)
    expect(rule.test.test('.png')).toEqual(true)
    expect(rule.test.test('.gif')).toEqual(true)
    expect(rule.test.test('.svg')).toEqual(true)
    expect(rule.test.test('.JPG')).toEqual(true)
    expect(rule.test.test('.JPEG')).toEqual(true)
    expect(rule.test.test('.PNG')).toEqual(true)
    expect(rule.test.test('.GIF')).toEqual(true)
    expect(rule.test.test('.SVG')).toEqual(true)
    expect(rule.test.test('.webp')).toEqual(false)
    expect(rule.test.test('.WEBP')).toEqual(false)
    expect(webpRule.test.test('.webp')).toEqual(true)
    expect(webpRule.test.test('.WEBP')).toEqual(true)

    expect(convartPathsToRelative(rule)).toMatchSnapshot()
    expect(convartPathsToRelative(webpRule)).toMatchSnapshot()
  })

  it('can disable image types', () => {
    const nuxtMockObject = nuxtMock(true, { handleImages: ['jpeg'] })
    const options = nuxtMockObject.init(nuxtOptimizedImages)

    expect(options.webPackConfig.module.rules).toHaveLength(2)

    const rule = options.webPackConfig.module.rules[0]

    expect(rule.test.test('.jpg')).toEqual(true)
    expect(rule.test.test('.jpeg')).toEqual(true)
    expect(rule.test.test('.png')).toEqual(false)
    expect(rule.test.test('.gif')).toEqual(false)
    expect(rule.test.test('.svg')).toEqual(false)
    expect(rule.test.test('.JPG')).toEqual(true)
    expect(rule.test.test('.JPEG')).toEqual(true)
    expect(rule.test.test('.PNG')).toEqual(false)
    expect(rule.test.test('.GIF')).toEqual(false)
    expect(rule.test.test('.SVG')).toEqual(false)
    expect(rule.test.test('.webp')).toEqual(false)
    expect(rule.test.test('.WEBP')).toEqual(false)

    expect(convartPathsToRelative(rule)).toMatchSnapshot()
  })

  it('can optimize all images on production', () => {
    const nuxtMockObject = nuxtMock(false)
    const options = nuxtMockObject.init(nuxtOptimizedImages)

    expect(options.webPackConfig.module.rules).toHaveLength(3)

    const rule = options.webPackConfig.module.rules[0]
    const webpRule = options.webPackConfig.module.rules[1]

    expect(convartPathsToRelative(rule)).toMatchSnapshot()
    expect(convartPathsToRelative(webpRule)).toMatchSnapshot()
  })
})
