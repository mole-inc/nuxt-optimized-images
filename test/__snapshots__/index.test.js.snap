// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nuxt-optimized-images can disable image types 1`] = `
Object {
  "oneOf": Array [
    Object {
      "resourceQuery": /\\(include\\.\\*original\\|original\\.\\*include\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/raw-loader/export-loader.js",
        },
        Object {
          "loader": "raw-loader",
        },
      ],
    },
    Object {
      "resourceQuery": /\\(inline\\.\\*original\\|original\\.\\*inline\\)/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": undefined,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /\\(url\\.\\*original\\|original\\.\\*url\\)/,
      "use": Array [
        Object {
          "loader": "file-loader",
          "options": Object {
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /url/,
      "use": Array [
        Object {
          "loader": "file-loader",
          "options": Object {
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /inline/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": undefined,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /include/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/raw-loader/export-loader.js",
        },
        Object {
          "loader": "raw-loader",
        },
      ],
    },
    Object {
      "resourceQuery": /original/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /lqip\\(&\\|\\$\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/lqip-loader/picture-export-loader.js",
        },
        Object {
          "loader": "@mole-inc/lqip-loader",
          "options": Object {
            "name": "[path][name].[ext]",
          },
        },
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /size/,
      "use": Array [
        Object {
          "loader": "responsive-loader",
          "options": Object {
            "adapter": [Function],
            "name": "[path][name]--[width].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /sqip\\(&\\|\\$\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/sqip-loader/export-loader.js",
        },
        Object {
          "loader": "sqip-loader",
          "options": Object {
            "name": "[path][name].[ext]",
          },
        },
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "[path][name].[ext]",
          },
        },
        Object {
          "loader": "img-loader",
          "options": Object {
            "plugins": Array [],
          },
        },
      ],
    },
  ],
  "test": /\\\\\\.\\(jpe\\?g\\)\\$/i,
}
`;

exports[`nuxt-optimized-images can optimize all images on production 1`] = `
Object {
  "oneOf": Array [
    Object {
      "resourceQuery": /\\(include\\.\\*original\\|original\\.\\*include\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/raw-loader/export-loader.js",
        },
        Object {
          "loader": "raw-loader",
        },
      ],
    },
    Object {
      "resourceQuery": /\\(inline\\.\\*original\\|original\\.\\*inline\\)/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": undefined,
            "name": "img/[contenthash:7].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /\\(url\\.\\*original\\|original\\.\\*url\\)/,
      "use": Array [
        Object {
          "loader": "file-loader",
          "options": Object {
            "name": "img/[contenthash:7].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /url/,
      "use": Array [
        Object {
          "loader": "file-loader",
          "options": Object {
            "name": "img/[contenthash:7].[ext]",
          },
        },
        Object {
          "loader": "img-loader",
          "options": Object {
            "plugins": Array [
              [Function],
              [Function],
              [Function],
              [Function],
            ],
          },
        },
      ],
    },
    Object {
      "resourceQuery": /inline/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": undefined,
            "name": "img/[contenthash:7].[ext]",
          },
        },
        Object {
          "loader": "img-loader",
          "options": Object {
            "plugins": Array [
              [Function],
              [Function],
              [Function],
              [Function],
            ],
          },
        },
      ],
    },
    Object {
      "resourceQuery": /include/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/raw-loader/export-loader.js",
        },
        Object {
          "loader": "raw-loader",
        },
        Object {
          "loader": "img-loader",
          "options": Object {
            "plugins": Array [
              [Function],
              [Function],
              [Function],
              [Function],
            ],
          },
        },
      ],
    },
    Object {
      "resourceQuery": /original/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "img/[contenthash:7].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /lqip\\(&\\|\\$\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/lqip-loader/picture-export-loader.js",
        },
        Object {
          "loader": "@mole-inc/lqip-loader",
          "options": Object {
            "name": "img/[contenthash:7].[ext]",
          },
        },
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "img/[contenthash:7].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /size/,
      "use": Array [
        Object {
          "loader": "responsive-loader",
          "options": Object {
            "adapter": [Function],
            "name": "img/[contenthash:7]-[width].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /sqip\\(&\\|\\$\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/sqip-loader/export-loader.js",
        },
        Object {
          "loader": "sqip-loader",
          "options": Object {
            "name": "img/[contenthash:7].[ext]",
          },
        },
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "img/[contenthash:7].[ext]",
          },
        },
        Object {
          "loader": "img-loader",
          "options": Object {
            "plugins": Array [
              [Function],
              [Function],
              [Function],
              [Function],
            ],
          },
        },
      ],
    },
    Object {
      "resourceQuery": /webp/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "mimetype": "image/webp",
            "name": "img/[contenthash:7].webp",
          },
        },
        Object {
          "loader": "@mole-inc/webp-loader",
          "options": Object {},
        },
      ],
    },
    Object {
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "img/[contenthash:7].[ext]",
          },
        },
        Object {
          "loader": "img-loader",
          "options": Object {
            "plugins": Array [
              [Function],
              [Function],
              [Function],
              [Function],
            ],
          },
        },
      ],
    },
  ],
  "test": /\\\\\\.\\(jpe\\?g\\|png\\|svg\\|gif\\)\\$/i,
}
`;

exports[`nuxt-optimized-images can optimize all images on production 2`] = `
Object {
  "oneOf": Array [
    Object {
      "resourceQuery": /\\(include\\.\\*original\\|original\\.\\*include\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/raw-loader/export-loader.js",
        },
        Object {
          "loader": "raw-loader",
        },
      ],
    },
    Object {
      "resourceQuery": /\\(inline\\.\\*original\\|original\\.\\*inline\\)/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": undefined,
            "name": "img/[contenthash:7].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /\\(url\\.\\*original\\|original\\.\\*url\\)/,
      "use": Array [
        Object {
          "loader": "file-loader",
          "options": Object {
            "name": "img/[contenthash:7].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /url/,
      "use": Array [
        Object {
          "loader": "file-loader",
          "options": Object {
            "name": "img/[contenthash:7].[ext]",
          },
        },
        Object {
          "loader": "@mole-inc/webp-loader",
          "options": Object {},
        },
      ],
    },
    Object {
      "resourceQuery": /inline/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": undefined,
            "name": "img/[contenthash:7].[ext]",
          },
        },
        Object {
          "loader": "@mole-inc/webp-loader",
          "options": Object {},
        },
      ],
    },
    Object {
      "resourceQuery": /include/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/raw-loader/export-loader.js",
        },
        Object {
          "loader": "raw-loader",
        },
        Object {
          "loader": "@mole-inc/webp-loader",
          "options": Object {},
        },
      ],
    },
    Object {
      "resourceQuery": /original/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "img/[contenthash:7].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /lqip\\(&\\|\\$\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/lqip-loader/picture-export-loader.js",
        },
        Object {
          "loader": "@mole-inc/lqip-loader",
          "options": Object {
            "name": "img/[contenthash:7].[ext]",
          },
        },
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "img/[contenthash:7].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /size/,
      "use": Array [
        Object {
          "loader": "responsive-loader",
          "options": Object {
            "adapter": undefined,
            "name": "img/[contenthash:7]-[width].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /sqip\\(&\\|\\$\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/sqip-loader/export-loader.js",
        },
        Object {
          "loader": "sqip-loader",
          "options": Object {
            "name": "img/[contenthash:7].[ext]",
          },
        },
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "img/[contenthash:7].[ext]",
          },
        },
        Object {
          "loader": "@mole-inc/webp-loader",
          "options": Object {},
        },
      ],
    },
    Object {
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "img/[contenthash:7].[ext]",
          },
        },
        Object {
          "loader": "@mole-inc/webp-loader",
          "options": Object {},
        },
      ],
    },
  ],
  "test": /\\\\\\.webp\\$/i,
}
`;

exports[`nuxt-optimized-images handles all images by default 1`] = `
Object {
  "oneOf": Array [
    Object {
      "resourceQuery": /\\(include\\.\\*original\\|original\\.\\*include\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/raw-loader/export-loader.js",
        },
        Object {
          "loader": "raw-loader",
        },
      ],
    },
    Object {
      "resourceQuery": /\\(inline\\.\\*original\\|original\\.\\*inline\\)/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": undefined,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /\\(url\\.\\*original\\|original\\.\\*url\\)/,
      "use": Array [
        Object {
          "loader": "file-loader",
          "options": Object {
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /url/,
      "use": Array [
        Object {
          "loader": "file-loader",
          "options": Object {
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /inline/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": undefined,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /include/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/raw-loader/export-loader.js",
        },
        Object {
          "loader": "raw-loader",
        },
      ],
    },
    Object {
      "resourceQuery": /original/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /lqip\\(&\\|\\$\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/lqip-loader/picture-export-loader.js",
        },
        Object {
          "loader": "@mole-inc/lqip-loader",
          "options": Object {
            "name": "[path][name].[ext]",
          },
        },
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /size/,
      "use": Array [
        Object {
          "loader": "responsive-loader",
          "options": Object {
            "adapter": [Function],
            "name": "[path][name]--[width].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /sqip\\(&\\|\\$\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/sqip-loader/export-loader.js",
        },
        Object {
          "loader": "sqip-loader",
          "options": Object {
            "name": "[path][name].[ext]",
          },
        },
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /webp/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "mimetype": "image/webp",
            "name": "[path][name].webp",
          },
        },
        Object {
          "loader": "@mole-inc/webp-loader",
          "options": Object {},
        },
      ],
    },
    Object {
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "[path][name].[ext]",
          },
        },
        Object {
          "loader": "img-loader",
          "options": Object {
            "plugins": Array [],
          },
        },
      ],
    },
  ],
  "test": /\\\\\\.\\(jpe\\?g\\|png\\|svg\\|gif\\)\\$/i,
}
`;

exports[`nuxt-optimized-images handles all images by default 2`] = `
Object {
  "oneOf": Array [
    Object {
      "resourceQuery": /\\(include\\.\\*original\\|original\\.\\*include\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/raw-loader/export-loader.js",
        },
        Object {
          "loader": "raw-loader",
        },
      ],
    },
    Object {
      "resourceQuery": /\\(inline\\.\\*original\\|original\\.\\*inline\\)/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": undefined,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /\\(url\\.\\*original\\|original\\.\\*url\\)/,
      "use": Array [
        Object {
          "loader": "file-loader",
          "options": Object {
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /url/,
      "use": Array [
        Object {
          "loader": "file-loader",
          "options": Object {
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /inline/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": undefined,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /include/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/raw-loader/export-loader.js",
        },
        Object {
          "loader": "raw-loader",
        },
      ],
    },
    Object {
      "resourceQuery": /original/,
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /lqip\\(&\\|\\$\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/lqip-loader/picture-export-loader.js",
        },
        Object {
          "loader": "@mole-inc/lqip-loader",
          "options": Object {
            "name": "[path][name].[ext]",
          },
        },
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /size/,
      "use": Array [
        Object {
          "loader": "responsive-loader",
          "options": Object {
            "adapter": undefined,
            "name": "[path][name]--[width].[ext]",
          },
        },
      ],
    },
    Object {
      "resourceQuery": /sqip\\(&\\|\\$\\)/,
      "use": Array [
        Object {
          "loader": "../lib/loaders/sqip-loader/export-loader.js",
        },
        Object {
          "loader": "sqip-loader",
          "options": Object {
            "name": "[path][name].[ext]",
          },
        },
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
    Object {
      "use": Array [
        Object {
          "loader": "url-loader",
          "options": Object {
            "fallback": "file-loader",
            "limit": 1000,
            "name": "[path][name].[ext]",
          },
        },
      ],
    },
  ],
  "test": /\\\\\\.webp\\$/i,
}
`;
